File: static/js/haversine.js
/**
* haversine.js: contains necessary global functions needed to compute the distance
*               between two GPS coordinate points (longitude, latitude).
*/

/**
 * rad: converts degrees to radians.
 */
  var rad = function(x) {
    return x * Math.PI / 180;
  };

/**
 * getDistance: computes distance between two GPS coordinate points.
 *
 * @radius, earth's mean radius (m)
 * @a, haversine equation
 * @c, haversine equation
 * @d, haversine equation, distance between two GPS coordinate points (m)
 */
  var getDistance = function(p1, p2) {
    var radius = 6378137;
    var dLat = rad(p2.lat() - p1.lat());
    var dLong = rad(p2.lng() - p1.lng());
    var a = Math.sin(dLat / 2) * Math.sin(dLat / 2) +
      Math.cos(rad(p1.lat())) * Math.cos(rad(p2.lat())) *
      Math.sin(dLong / 2) * Math.sin(dLong / 2);
    var c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
    var d = radius * c;
    return d;
  };

File: package/haversine.py
#!/usr/bin/python

## @haversine.py
#  This file contains the necesary methods needed to compute the distance between
#      two GPS coordinate points (longitude, latitude).
from math import radians, cos, sin, asin, sqrt

## get_distance: computes the distance between two GPS coordinate points using the
#                haversine formula.
#
#  Note: this function returns the haversine distance in meters.
def get_distance(p1, p2):
  # radius of the earth (meters)
  radius = 6378137

  # convert decimal degrees to radians
  lon1, lat1, lon2, lat2 = map(radians, [p1['longitude'], p1['latitude'], p2['longitude'], p2['latitude']])

  # haversine formula
  dlon = lon2 - lon1
  dlat = lat2 - lat1
  a = sin(dlat/2)**2 + cos(lat1) * cos(lat2) * sin(dlon/2)**2
  c = 2 * asin(sqrt(a))

  # return distance
  d = radius * c
  return d
